### ISSUE TITLE
klp_tc_10 failing on s390x?

### ISSUE THREAD

**Issue by joe-lawrence** [lpechacek/qa_test_klp#17]
We noticed that klp_tc_10 is failing in our s390x tests (both with modified el9 kernels and upstream v5.18):
```
[08:09:20] *** Check that live patch is blocked
[08:09:22] TEST CASE ABORT patching finished prematurely
[08:09:22] *** Removing patches
[08:09:22] *** Disabling and removing module klp_tc_10_livepatch
[08:09:24] *** Removing module klp_test_support_mod
[08:09:24] rmmod: ERROR: Module klp_test_support_mod is in use
```
Commenting out the interesting parts of the test (the graph tracing) will allow the test to pass:
```
    klp_tc_milestone "Enable graph tracing for orig_do_sleep"
    echo "orig_do_sleep" > /sys/kernel/debug/tracing/set_ftrace_filter
    echo "function_graph" > /sys/kernel/debug/tracing/current_tracer
```
If I read the code correctly, for this test:
1. A module defines function A() to call its function B()
2. ftrace graph tracing is set on ~A()~ B()
3. Someone calls A() so it invokes B()
4. The module from (1) ~B()~ A() is patched by a livepatch
5. The livepatch transition should block until (3) has completed and A() and B() are off the call stack

That seems like a reasonable test and I would have expected to confirm step (5).  Is there anything special about s390x w/respect to klp_tc_10 or the consistency model?

edit: corrected the above steps as @mirab highlighed, where A() is orig_sleep_uninterruptable_set() and B() is orig_do_sleep()

**Comment by mirab** [comment]
Hi,

thanks for the report.

I do not think there is anything special regarding the consistency model and the test on s390x. Its purpose is to ensure that the unwinding works correctly through the function graph tracing if I am not mistaken (the tracing is set on B() while A() is patched by a livepatch).

So there may be a problem with the unwinding. Debugging time...

**Comment by marcosps** [comment]
I managed to reproduce the problem in a s390 VM. I will add more data here as soon as find something useful.

**Comment by marcosps** [comment]
@joe-lawrence I tried to reproduce the issue here, but I had issues while running a kernel on s390 from 5.18 onwards. I bisected the problematic:

  commit 30de14b1884ba609fc1acfba5b40309e3a6ccefe
  Author: Sven Schnelle <svens@linux.ibm.com>
  Date:   Fri Apr 8 14:51:26 2022 +0200
  
     s390: current_stack_pointer shouldn't be a function

We could boot a VM correctly when this commit was reverted. But a few days ago the commit was reverted by e3c11025bcd2142a61abe5806b2f86a0e78118df:

s390: avoid using global register for current_stack_pointer

With the fix in place, or the problematic commit reverted, the klp_tc_10 was running successfully in our tests. Would you mind rerunning the test with e3c11025bcd2142a61abe5806b2f86a0e78118df applied? Thanks!

**Comment by joe-lawrence** [comment]
@marcosps : sorry for belated response, it's a holiday weekend here :turkey:  :beer:.  I did use the time to slowly work through my own bisect (always 12+ steps, always :laughing:)  and while I didn't run into any issue booting with or without 30de14b1884b ("s390: current_stack_pointer shouldn't be a function"), the bisect finally landed on:
```
commit ded466e1806686794b403ebf031133bbaca76bb2
Author: Sumanth Korikkar <sumanthk@linux.ibm.com>
Date:   Tue Jul 26 18:57:59 2022 +0200

    s390/unwind: fix fgraph return address recovery
    
    When HAVE_FUNCTION_GRAPH_RET_ADDR_PTR is defined, the return
    address to the fgraph caller is recovered by tagging it along with the
    stack pointer of ftrace stack. This makes the stack unwinding more
    reliable.
    
    When the fgraph return address is modified to return_to_handler,
    ftrace_graph_ret_addr tries to restore it to the original
    value using tagged stack pointer.
    
    Fix this by passing tagged sp to ftrace_graph_ret_addr.
    
    Fixes: d81675b60d09 ("s390/unwind: recover kretprobe modified return address in stacktrace")
    Cc: <stable@vger.kernel.org> # 5.18
    Reviewed-by: Vasily Gorbik <gor@linux.ibm.com>
    Signed-off-by: Sumanth Korikkar <sumanthk@linux.ibm.com>
    Signed-off-by: Alexander Gordeev <agordeev@linux.ibm.com>

 arch/s390/include/asm/unwind.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
```
as the commit that fixes this issue.  I cherry-picked it back to v5.18 and observed the test is now successful.

**Comment by mirab** [comment]
Thanks for the update, Joe. Let me close the issue then.
